### ISSUE TITLE
AtomTypeError

### ISSUE THREAD

**Issue by Pegah-B** [ReactionMechanismGenerator/RMG-Py#2289]
### Topic
General area which your question is related to.
 - [ ] Installation of RMG
 - [x] Running an RMG job
 - [ ] Using RMG API
 - [ ] Arkane (formerly CanTherm)
 - [ ] Dependencies
 - [x] An error message 

### Context
I wanted to construct a chemical reaction network started from pyruvate, glyoxylate, bicarbonate buffer dissolved in water and in presence of nitrogen. Here is the RMG input file I am trying to run:

```
#%% Data sources
database(
    thermoLibraries = ['primaryThermoLibrary'],
    reactionLibraries = [],
    seedMechanisms = [],
    kineticsDepositories = ['training'],
    kineticsFamilies = 'default',
    kineticsEstimator = 'rate rules'
)
#%% Species relevant:

species(
    label = 'water', #H2O
    reactive = True,
    structure = SMILES("O"),
)
species(
    label = 'glyoxylate',  #C2HO3-
    reactive = True,
    structure = SMILES('C(=O)C(=O)[O]')
)
species(
    label = 'pyruvate',  #C3H3O3-
    reactive = True,
    structure = SMILES('CC(=O)C(=O)[O]')
)

species(
    label = 'bicarbonate',  #HCO3- 
    reactive = True,
    structure = SMILES('OC([O])=O')
)

species(
    label = 'nitrogen',  #N2
    reactive = True,
    structure = SMILES('N#N')
)

#%% Liquid state reactions:    
liquidReactor(
    temperature = (300,'K'),
    initialConcentrations = {
            "water" : 1000,
            "pyruvate" : 200,
            "glyoxylate" : 600,
            "bicarbonate" :500,
            "nitrogen" : 500,
            },
    terminationTime = (1e6,'s'),   
)
solvation('water')

#%% Speed options:
model(
    toleranceKeepInEdge=0,
    toleranceMoveToCore = 0.1,
    toleranceInterruptSimulation = 0.1,
    maximumEdgeSpecies = 100000,
    filterReactions = True
      )
#%% Simulator options:
simulator(
    atol=1e-16,
    rtol=1e-8,
    )
#%% Saving options:
options(
    units='si',
    generateOutputHTML=False,
    generatePlots=False,
    saveEdgeSpecies=False,
    saveSimulationProfiles=False,
    generateSeedEachIteration=True,
)
```

and here is part of the log file including an error I am facing:

```
.
.
.

After model enlargement:
    The model core has 21 species and 14 reactions
    The model edge has 2792 species and 3930 reactions

Saving current model core to Chemkin file...
Chemkin file contains 14 reactions.
Saving annotated version of Chemkin file...
Chemkin file contains 14 reactions.
Updating RMG execution statistics...
    Execution time (DD:HH:MM:SS): 00:00:50:20
    Memory used: 954.71 MB

Making seed mechanism...
Conducting simulation of reaction system 1...
At time 1.6529e-11 s, species [C-]#[O+](8) at 0.10045677834812579 exceeded the minimum rate for simulation interruption of 0.1
Reached max number of objects...preparing to terminate
At time 1.6529e-11 s, species [C-]#[O+](8) at rate ratio 0.10045677834812579 exceeded the minimum rate for moving to model core of 0.1
terminating simulation due to interrupt...

Adding species [C-]#[O+](8) to model core

Summary of Model Enlargement
---------------------------------
Added 1 new core species
    [C-]#[O+](8)
Created 0 new edge species
Moved 3 reactions from edge to core
    [CH]=O(17) + bicarbonate(4) <=> [C-]#[O+](8) + O=C(O)O(51)
    OH(D)(6) + [CH]=O(17) <=> water(1) + [C-]#[O+](8)
    water(1) + [C-]#[O+](8) <=> O=CO(652)
Added 0 new core reactions
Created 0 new edge reactions

After model enlargement:
    The model core has 22 species and 17 reactions
    The model edge has 2791 species and 3927 reactions

At time 1.0000e+06 s, reached target termination time.

For reaction generation 1 process is used.
Generating thermo for new species...
Error: Could not update atomtypes for this molecule:
multiplicity 1
1  O u0 p2 c0 {2,S} {8,S}
2  O u0 p2 c0 {1,S} {9,S}
3  O u0 p2 c0 {4,S} {9,S}
4  O u0 p1 c+1 {3,S} {11,D}
5  O u0 p2 c0 {8,D}
6  O u0 p2 c0 {9,D}
7  O u0 p2 c0 {10,D}
8  C u0 p0 c0 {1,S} {5,D} {10,S}
9  C u0 p0 c0 {2,S} {3,S} {6,D}
10 C u0 p0 c0 {7,D} {8,S} {12,S}
11 C u2 p0 c-1 {4,D} {13,S}
12 H u0 p0 c0 {10,S}
13 H u0 p0 c0 {11,S}

Traceback (most recent call last):
  File "/opt/anaconda3/envs/rmg_env/bin/rmg.py", line 118, in <module>
    main()
  File "/opt/anaconda3/envs/rmg_env/bin/rmg.py", line 112, in main
    rmg.execute(**kwargs)
  File "/opt/anaconda3/envs/rmg_env/lib/python3.7/site-packages/rmgpy/rmg/main.py", line 923, in execute
    trimolecular_react=self.trimolecular_react)
  File "/opt/anaconda3/envs/rmg_env/lib/python3.7/site-packages/rmgpy/rmg/model.py", line 614, in enlarge
    self.apply_thermo_to_species(procnum)
  File "/opt/anaconda3/envs/rmg_env/lib/python3.7/site-packages/rmgpy/rmg/model.py", line 824, in apply_thermo_to_species
    self.generate_thermo(spc, rename=True)
  File "/opt/anaconda3/envs/rmg_env/lib/python3.7/site-packages/rmgpy/rmg/model.py", line 831, in generate_thermo
    submit(spc, self.solvent_name)
  File "/opt/anaconda3/envs/rmg_env/lib/python3.7/site-packages/rmgpy/thermo/thermoengine.py", line 175, in submit
    spc.thermo = evaluator(spc, solvent_name=solvent_name)
  File "/opt/anaconda3/envs/rmg_env/lib/python3.7/site-packages/rmgpy/thermo/thermoengine.py", line 160, in evaluator
    thermo = generate_thermo_data(spc, solvent_name=solvent_name)
  File "/opt/anaconda3/envs/rmg_env/lib/python3.7/site-packages/rmgpy/thermo/thermoengine.py", line 143, in generate_thermo_data
    return process_thermo_data(spc, thermo0, thermo_class, solvent_name)
  File "/opt/anaconda3/envs/rmg_env/lib/python3.7/site-packages/rmgpy/thermo/thermoengine.py", line 66, in process_thermo_data
    solute_data = solvation_database.get_solute_data(spc)
  File "/opt/anaconda3/envs/rmg_env/lib/python3.7/site-packages/rmgpy/data/solvation.py", line 694, in get_solute_data
    solute_data = self.get_solute_data_from_groups(species)
  File "/opt/anaconda3/envs/rmg_env/lib/python3.7/site-packages/rmgpy/data/solvation.py", line 751, in get_solute_data_from_groups
    solute_data = self.estimate_solute_via_group_additivity(molecule)
  File "/opt/anaconda3/envs/rmg_env/lib/python3.7/site-packages/rmgpy/data/solvation.py", line 861, in estimate_solute_via_group_additivity
    saturated_struct, added_to_pairs = self.transform_lone_pairs(saturated_struct)
  File "/opt/anaconda3/envs/rmg_env/lib/python3.7/site-packages/rmgpy/data/solvation.py", line 796, in transform_lone_pairs
    saturated_struct.update()
  File "rmgpy/molecule/molecule.py", line 1137, in rmgpy.molecule.molecule.Molecule.update
  File "rmgpy/molecule/molecule.py", line 1362, in rmgpy.molecule.molecule.Molecule.update_atomtypes
  File "rmgpy/molecule/molecule.py", line 1357, in rmgpy.molecule.molecule.Molecule.update_atomtypes
  File "rmgpy/molecule/atomtype.py", line 813, in rmgpy.molecule.atomtype.get_atomtype
  File "rmgpy/molecule/atomtype.py", line 841, in rmgpy.molecule.atomtype.get_atomtype
rmgpy.exceptions.AtomTypeError: Unable to determine atom type for atom C..-, which has 1 single bonds, 1 double bonds (1 to O, 0 to S, 0 others), 0 triple bonds, 0 quadruple bonds, 0 benzene bonds, 0 lone pairs, and -1 charge.
```

How should I fix the error and have the model converged? 

### Installation Information
Describe your installation method and system information if applicable.
 - OS (include version if known): macOS 10.14.6
 - Installation method: installation from binary , with anaconda
 - RMG version information:
   
   - RMG-Py: 3.1.0
   - RMG-database: 3.1.0

**Comment by xiaoruiDong** [comment]
HI @Pegah-B, Thank you for being interested in using RMG-Py. Due to RMG's limitation, it can only work properly for neutral species chemistry and some catalytic reactions. I don't think it can generate useful results for your case right now. (your input is all like anions, right? BTW, it is `C(=O)C(=O)[O-]`, `CC(=O)C(=O)[O-]`, `OC([O-])=O`for the anion SMILES). RMG doesn't have the necessary atom types as well as reaction families for anions. There is a long-term plan of adding them, but the functionality won't be available in one or two years. Let me know if you have other questions.

**Comment by Pegah-B** [comment]
Thanks for your reply. Based on what you said, I switched to a  totally different scenarios and tried to start with initial neutral species this time. However, I still get an error message about an unknown created species, now just after initialization complete, and I do not know why! I would really appreciate it if you help me to figure this out. Here is my input file:

```
#%% Data sources
database(
    thermoLibraries = ['primaryThermoLibrary'],
    reactionLibraries = [],
    seedMechanisms = [],
    kineticsDepositories = ['training'],
    kineticsFamilies = 'default',
    kineticsEstimator = 'rate rules'
)

#%% Species relevant:
species(
    label = 'water', #H2O
    reactive = True,
    structure = SMILES("O"),
)

species(
    label = 'ammonia',  #NH3
    reactive = True,
    structure = SMILES('N')
)

species(
label = 'methane',  #CH4
reactive = True,
structure = SMILES("C")
)

species(
    label = 'hydrogen cyanide',  #HCN
    reactive = True,
    structure = SMILES('C#N')
)

species(
    label = 'hydrogen sulphide',  #H2S
    reactive = True,
    structure = SMILES('S')
)

species(
    label = 'nitrogen', #N2
    reactive = True,
    structure = SMILES("N#N"),
)


#%% Liquid state reactions:    

liquidReactor(
    temperature = (300,'K'),
    initialConcentrations = {
            "water" : 3,
            "ammonia" : 1,
            "methane": 1,
            "nitrogen" : 1,
            "hydrogen cyanide" :1,
            "hydrogen sulphide" : 1,
            },
    terminationTime = (1000,'s'),   
)
solvation('water')

#%% Speed options:
model(
    toleranceKeepInEdge=0,
    toleranceMoveToCore = 0.1,
    toleranceInterruptSimulation = 0.1,
    maximumEdgeSpecies = 100000,
    #filterReactions = True
      )
#%% Simulator options:
simulator(
    atol=1e-16,
    rtol=1e-8,
    )
#%% Saving options:
options(
    units='si',
    generateOutputHTML=True,
    generatePlots=False,
    saveEdgeSpecies=False,
    saveSimulationProfiles=False,
    generateSeedEachIteration=True,
)
```

And here is the log file:

```
Loading transport library from GRI-Mech.py in /opt/anaconda3/envs/rmg_env/share/rmgdatabase/transport/libraries...
Loading transport library from OneDMinN2.py in /opt/anaconda3/envs/rmg_env/share/rmgdatabase/transport/libraries...
Loading transport library from NOx2018.py in /opt/anaconda3/envs/rmg_env/share/rmgdatabase/transport/libraries...
Loading transport library from PrimaryTransportLibrary.py in /opt/anaconda3/envs/rmg_env/share/rmgdatabase/transport/libraries...
Loading transport group database from /opt/anaconda3/envs/rmg_env/share/rmgdatabase/transport/groups...
Loading frequencies group database from /opt/anaconda3/envs/rmg_env/share/rmgdatabase/statmech/groups...
Loading Platts additivity group database from /opt/anaconda3/envs/rmg_env/share/rmgdatabase/solvation/groups...
Loading thermodynamics library from primaryThermoLibrary.py in /opt/anaconda3/envs/rmg_env/share/rmgdatabase/thermo/libraries...
Loading thermodynamics group database from /opt/anaconda3/envs/rmg_env/share/rmgdatabase/thermo/groups...
Adding rate rules from training set in kinetics families...
Warning: Cphigh is above the theoretical CpInf value for ThermoData object
ThermoData(Tdata=([300,400,500,600,800,1000,1500],'K'), Cpdata=([36.6869,38.2195,39.253,40.4473,45.0357,50.6323,54.2027],'J/(mol*K)'), H298=(310.047,'kJ/mol'), S298=(162.808,'J/(mol*K)'), Cp0=(29.1007,'J/(mol*K)'), CpInf=(37.4151,'J/(mol*K)'), comment="""Thermo group additivity estimation: group(O2s-CsCs) + group(CsJ2_singlet-CsH)""").
The thermo for this species is probably wrong! Setting CpInf = Cphigh for Entropy calculationat T = 2000.0 K...
Warning: Cphigh is above the theoretical CpInf value for ThermoData object
ThermoData(Tdata=([300,400,500,600,800,1000,1500],'K'), Cpdata=([36.6869,38.2195,39.253,40.4473,45.0357,50.6323,54.2027],'J/(mol*K)'), H298=(310.047,'kJ/mol'), S298=(162.808,'J/(mol*K)'), Cp0=(29.1007,'J/(mol*K)'), CpInf=(37.4151,'J/(mol*K)'), comment="""Thermo group additivity estimation: group(O2s-CsCs) + group(CsJ2_singlet-CsH)""").
The thermo for this species is probably wrong! Setting CpInf = Cphigh for Entropy calculationat T = 1666.6666666666665 K...
Warning: Cphigh is above the theoretical CpInf value for ThermoData object
ThermoData(Tdata=([300,400,500,600,800,1000,1500],'K'), Cpdata=([36.6869,38.2195,39.253,40.4473,45.0357,50.6323,54.2027],'J/(mol*K)'), H298=(310.047,'kJ/mol'), S298=(162.808,'J/(mol*K)'), Cp0=(29.1007,'J/(mol*K)'), CpInf=(37.4151,'J/(mol*K)'), comment="""Thermo group additivity estimation: group(O2s-CsCs) + group(CsJ2_singlet-CsH)""").
The thermo for this species is probably wrong! Setting CpInf = Cphigh for Entropy calculationat T = 2000.0 K...
Warning: Cphigh is above the theoretical CpInf value for ThermoData object
ThermoData(Tdata=([300,400,500,600,800,1000,1500],'K'), Cpdata=([36.6869,38.2195,39.253,40.4473,45.0357,50.6323,54.2027],'J/(mol*K)'), H298=(310.047,'kJ/mol'), S298=(162.808,'J/(mol*K)'), Cp0=(29.1007,'J/(mol*K)'), CpInf=(37.4151,'J/(mol*K)'), comment="""Thermo group additivity estimation: group(O2s-CsCs) + group(CsJ2_singlet-CsH)""").
The thermo for this species is probably wrong! Setting CpInf = Cphigh for Entropy calculationat T = 1666.6666666666665 K...
Filling in rate rules in kinetics families by averaging...
Enabling diffusion-limited kinetics...
Setting solvent data for water
NOT generating reactions for unreactive species Ar

Summary of Model Enlargement
---------------------------------
Added 1 new core species
    Ar
Created 0 new edge species
Added 0 new core reactions
Created 0 new edge reactions

After model enlargement:
    The model core has 1 species and 0 reactions
    The model edge has 6 species and 0 reactions


NOT generating reactions for unreactive species He

Summary of Model Enlargement
---------------------------------
Added 1 new core species
    He
Created 0 new edge species
Added 0 new core reactions
Created 0 new edge reactions

After model enlargement:
    The model core has 2 species and 0 reactions
    The model edge has 6 species and 0 reactions


NOT generating reactions for unreactive species Ne

Summary of Model Enlargement
---------------------------------
Added 1 new core species
    Ne
Created 0 new edge species
Added 0 new core reactions
Created 0 new edge reactions

After model enlargement:
    The model core has 3 species and 0 reactions
    The model edge has 6 species and 0 reactions


Adding species water(1) to model core

Summary of Model Enlargement
---------------------------------
Added 1 new core species
    water(1)
Created 0 new edge species
Added 0 new core reactions
Created 0 new edge reactions

After model enlargement:
    The model core has 4 species and 0 reactions
    The model edge has 5 species and 0 reactions


Adding species ammonia(2) to model core

Summary of Model Enlargement
---------------------------------
Added 1 new core species
    ammonia(2)
Created 0 new edge species
Added 0 new core reactions
Created 0 new edge reactions

After model enlargement:
    The model core has 5 species and 0 reactions
    The model edge has 4 species and 0 reactions


Adding species methane(3) to model core

Summary of Model Enlargement
---------------------------------
Added 1 new core species
    methane(3)
Created 0 new edge species
Added 0 new core reactions
Created 0 new edge reactions

After model enlargement:
    The model core has 6 species and 0 reactions
    The model edge has 3 species and 0 reactions


Adding species hydrogen cyanide(4) to model core

Summary of Model Enlargement
---------------------------------
Added 1 new core species
    hydrogen cyanide(4)
Created 0 new edge species
Added 0 new core reactions
Created 0 new edge reactions

After model enlargement:
    The model core has 7 species and 0 reactions
    The model edge has 2 species and 0 reactions


Adding species hydrogen sulphide(5) to model core

Summary of Model Enlargement
---------------------------------
Added 1 new core species
    hydrogen sulphide(5)
Created 0 new edge species
Added 0 new core reactions
Created 0 new edge reactions

After model enlargement:
    The model core has 8 species and 0 reactions
    The model edge has 1 species and 0 reactions


Adding species nitrogen(6) to model core

Summary of Model Enlargement
---------------------------------
Added 1 new core species
    nitrogen(6)
Created 0 new edge species
Added 0 new core reactions
Created 0 new edge reactions

After model enlargement:
    The model core has 9 species and 0 reactions
    The model edge has 0 species and 0 reactions


Initialization complete. Starting model generation.

Generating initial reactions...
For reaction generation 1 process is used.
Generating thermo for new species...
Species [OH] renamed OH(D) based on thermo library name
Species [H] renamed H based on thermo library name
Species [NH2] renamed NH2(D) based on thermo library name
Species [CH2] renamed CH2(S) based on thermo library name
Species [H][H] renamed H2 based on thermo library name
Error: Could not update atomtypes for this molecule:
multiplicity 1
1 N u0 p1 c0 {2,S} {4,S} {5,S}
2 N u0 p0 c+1 {1,S} {3,D} {6,S}
3 N u2 p1 c-1 {2,D}
4 H u0 p0 c0 {1,S}
5 H u0 p0 c0 {1,S}
6 H u0 p0 c0 {2,S}
```

**Comment by xiaoruiDong** [comment]
Thank you for letting us know about the issue. 

Apparently, the error is due to RMG being lack of knowledge of several atom types. Atom No.3 in this case, where N has two unpaired electrons. According to my tuition, Atom No.3 should be `N u0 p2 c-1 {2,D}` instead. I don't know how this molecule is generated during your RMG run, but it looks like there is probably something wrong when applying reaction templates to Nitrogenated species or generating resonance structures. 

I personally don't have much experience dealing with nitrogenated species, and I don't know someone is actively working on Nitrogen chemistry. @alongd, maybe you or one of your students can take a look at this case? 

I guess this species may be from N2 + NH3. I am not sure what actual chemistry you are trying to predict, but you can set N2 to non-reactive to temporarily work around this issue by 

```
species(
    label = 'nitrogen', #N2
    reactive = True,
    structure = SMILES("N#N"),
)
```

**Comment by Pegah-B** [comment]
Thanks, I tried it but I still faced the issue.  
Even reducing the above file to include only two species "HCN" and "H2O", I still got the following atomtype error:
```
Error: Could not update atomtypes for this molecule:
multiplicity 3
1 N u1 p0 c+1 {3,S} {4,D}
2 N u3 p1 c-1 {3,S}
3 C u0 p0 c0 {1,S} {2,S} {4,S} {5,S}
4 C u0 p0 c0 {1,D} {3,S} {6,S}
5 H u0 p0 c0 {3,S}
6 H u0 p0 c0 {4,S}
```

Could you please somebody help me with how to fix the atomtype error?? 
I would really appreciate it.

**Comment by alongd** [comment]
Hi @Pegah-B,
I ran the latest input file you posted, and the job did successfully complete. It didn't generate a large enough model, but that's something else to look at.

Could you please make sure you're running on the most updated RMG-Py and RMG-database branches (do `git pull` when on the `main` branch of each repo, then do `make` under the `RMG-Py` folder)?

**Comment by Pegah-B** [comment]
Thanks, the issue got fixed. I have already installed the latest version 3.1.0 both on mac and linux, through binary installation using Anaconda just according to the instruction http://reactionmechanismgenerator.github.io/RMG-Py/users/rmg/installation/anacondaUser.html. With that RMG, I kept getting an atomtype error, like reported above. There must something wrong with these kinds of inputs there. After your comment, I entirely reinstalled RMG 3.1.0 this time through source using Anaconda environment according to the updated instruction http://reactionmechanismgenerator.github.io/RMG-Py/users/rmg/installation/anacondaDeveloper.html. Now, everything works well.
